axelrod/strategies/axelrod_second.py

```diff
@@ -975,3 +975,329 @@ def strategy(self, opponent: Player) -> Action:
+        self.burned = False
+
+        self.defect_streak = 0
+        self.parity_streak = [0, 0]  # Counters that get (almost) alternatively incremented.
+        self.parity_bit = 0  # Which parity_streak to increment
+        self.parity_limit = 5  # When a parity streak hits this limit, alter strategy.
+        self.parity_hits = 0  # Counts how many times a parity_limit was hit.
+        # After hitting parity_hits 8 times, lower parity_limit to 3.
+
+    def try_return(self, to_return, lower_flags=True, inc_parity=False):
+        """
+        This will return to_return, with some end-of-turn logic.
+        """
+
+        if lower_flags and to_return == C:
+            # In most cases when Cooperating, we want to reduce the number that
+            # are scheduled.
+            self.more_coop -= 1
+            self.last_generous_n_turns_ago += 1
+
+        if inc_parity and to_return == D:
+            # In some cases we increment the `parity_streak` that we're on when
+            # we return a Defection.  In detect_parity_streak, `parity_streak`
+            # counts opponent's Defections.
+            self.parity_streak[self.parity_bit] += 1
+
+        return to_return
+
+    def detect_random(self, turn):
+        """
+        Calculates a modified Pearson's Chi Squared statistic on self.history,
+        and returns True (is random) if and only if the statistic is less than
+        or equal to 3.
+
+        Pearson's Chi Squared statistic = sum[ (E_i-O_i)^2 / E_i ], where O_i
+        are the observed matrix values, and E_i is calculated as number (of
+        defects) in the row times the number in the column over (total number
+        in the matrix minus 1).
+
+        We say this is modified because it differs from a usual Chi-Squared
+        test in that:
+
+        - Terms where expected counts are less than 1 get excluded.
+        - There's a check at the beginning on the first cell of the matrix.
+        - There's a check at the beginning for the recorded number of defects.
+
+        """
+        denom = turn - 2
+
+        if self.move_history[0, 0] / denom >= 0.8:
+            return False
+        if self.recorded_defects / denom < 0.25 or self.recorded_defects / denom > 0.75:
+            return False
+
+        # In each cell, we expect (for an independent distribution) the total
+        # number of recorded turns times the portion in that row times the
+        # portion in that column
+        expected_matrix = np.outer(self.move_history.sum(axis=1),
+                                   self.move_history.sum(axis=0)) / denom
+
+        chi_squared = 0.0
+        for i in range(4):
+            for j in range(2):
+                expect = expected_matrix[i, j]
+                if expect > 1.0:
+                    chi_squared += (expect - self.move_history[i, j]) ** 2 / expect
+
+        # Caching value only for testing purposes, not used otherwise
+        self.chi_squared = round(chi_squared, 3)
```

Comment:
So it doesn't look like you've gone with the suggestion of making the `chi_squared` calculation it's own function which could be tested?

You could essentially keep this exactly as it is:

```
+ def calculate_chi_squared(self):
+        expected_matrix = np.outer(self.move_history.sum(axis=1),
 +                                   self.move_history.sum(axis=0)) / denom
 +
 +        chi_squared = 0.0
 +        for i in range(4):
 +            for j in range(2):
 +                expect = expected_matrix[i, j]
 +                if expect > 1.0:
 +                    chi_squared += (expect - self.move_history[i, j]) ** 2 / expect
 +
 +        # Caching value only for testing purposes, not used otherwise
 +        return chi_squared
```

Then in `detect_random` you just use `self.calculate_chi_squared` but then, for the test, you can directly run a match for your strategy and call the `self.calculate_chi_squared` method to ensure it gives the correct value (with no need for the `self.chi_squared`.

My other suggestion was to have this as a completely static method/function but it's also fine to leave it as a method of the class.

Reply:
I didn't make the change, because it sounded like it was optional AND because I'm still not seeing how, under this alternative approach, I would test that `move_history` is getting filled out right.  If I understand your comment, it was that, this doesn't matter because I can test the Chi-Squared logic on my matrix by manually entering the matrix:

self.assertEqual(round(axelrod_second.calculate_chi_squared(np.matrix([[0, 2], [5, 6], [3, 6], [4, 2]])), 3), 2.395)

While I agree that this would successfully test the chi-squared logic.  How do I test that `move_history` correctly incremented to [[0, 2], [5, 6], [3, 6], [4, 2]] ?

Reply:
I'm sure I see the connection between testing the `move_history` and `calculate_chi_squared`. As far as I can tell you're not testing that in the tests anyway:

```
+        actions += [(D, C)]
+        self.versus_test(axelrod.Random(0.5), expected_actions=actions, seed=10, attrs={"chi_squared": 2.395})
+        # The history matrix will be [[0, 2], [5, 6], [3, 6], [4, 2]]
```

You can keep this as a sub method of the class if you prefer in which case nothing here changes except you no longer need to artificially set `self.calculate_chi_squared` as an attribute (which is rounded so not actually an accurate name).

So that test (which if I understand correctly is the only purpose of this rounded chi_squared attribute which we'd rather not need) would be modified to be an actual match. Something like (I haven't run this):

```
axl.seed(10)
player = self.player()
match = axl.Match((player, axl.Random()), turns=len(expected_actions))
actions = match.play()
self.assertEqual(actions, expected_actions)  # Just to be consistant with the correct test.
self.assertEqual(round(player.calculate_chi_squared()), 2.395)
```

Reply:
> I'm not sure I see the connection between testing the move_history and calculate_chi_squared.

calculate_chi_squared uses move_history.  You can get 2.935 precisely if move_history is filled out (correctly) as [[0, 2], [5, 6], [3, 6], [4, 2]]

Reply:
I think I understand better now...

Reply:
> self.assertEqual(round(player.calculate_chi_squared()), 2.395)

I wasn't making the connection that the player would still be in the _end-of-turn-30_ state after you ran the axl.Match(), and I was trying to figure out how I could check everything in the Match() function, because I was imagining that the player would disappear afterwards.  [I guess the way I wrote it, it would disappear.]

I'll change it like you have it.  Thanks.

Reply:
> calculate_chi_squared uses move_history. You can get 2.935 precisely if move_history is filled out (correctly) as [[0, 2], [5, 6], [3, 6], [4, 2]]

Cool, so we can either:

1. create `compute_chi_squared` as it's own function (outside the class) that in the test gets passed `np.array([[0, 2], [5, 6], [3, 6], [4, 2]])` as well as `recorded_defects` (which I believe it also uses). So `compute_chi_squared` would be a function that takes two arguments which is in turn called by `detect_random`. Something like:
   
   ```python
   def detect_random(self):
        chi_squared = compute_chi_squared(self.move_history, self.recorded_defects)
        if chi_squared > 3 ...
   ```
   The test would be:

    ```python
     self.assertEqual(round(axelrod_second.calculate_chi_squared(np.matrix([[0, 2], [5, 6], [3, 6], [4, 2]])),  ?), 2.395)      # Replace ? with the correct recorded_defects :)
     ```
   
2. Create a method which means we then need to test using a match.

I think option 1 is probably the cleanest way of doing this but 2 would do the trick also :)

Reply:
> I think I understand better now...

Cool :+1: